add a transport to support connecting with arbitrary streams

jhgaylor opened the pull request:
Hey there! Another PR to confuse you 😄 

I know this doesn't conform to the spec but I ended up needing this so I can spawn dart mcp servers in my dart application without going around the world and spawning a new process.

Is this something you might want to bring in? 

```
final server = mcp_dart.McpServer(...);

# create streams
final serverInputStreamController = StreamController<List<int>>();
final serverOutputStreamController = StreamController<List<int>>();

# attach the streams to the server transport
final serverTransport = IOStreamServerTransport(
  stream: serverInputStreamController.stream,
  sink: serverOutputStreamController.sink,
);
await server.connect(serverTransport);

# setup a client transport to use the same streams the server is connected to
clientTransport = IOStreamClientTransport(
  stream: serverOutputStreamController.stream,
  sink: serverInputStreamController.sink,
);
```

leehack commented:
That's an interesting approach. You could probably clean it up for the functions or variable names that include `stdio`.

Also, avoiding code duplication, we could also use this as a base while other transporters extends this. What do you think?

It would also be great if you could add some testing.

jhgaylor commented:
Absolutely. If you're interested in the code getting in I will do some work to clean it up and test it.

leehack commented:
Yep. Let's put that in.

leehack commented:
For this transporter implementation, I think there might be no difference between the server and the client. Basically, it's similar to https://github.com/modelcontextprotocol/typescript-sdk/blob/main/src/inMemory.ts, right?

jhgaylor commented:
I will get it renamed. It does look like both transports are the same. Where do you want them in the code? it looks like we dont currently have a shared module.dart. Would it be appropriate to put the inmemory transport there and export it?

jhgaylor commented:
This does seem a bit different from their in memory transport but we could build an inmemory transport on top of this one pretty easily I think

jhgaylor commented:
I am happy to make changes still, including renaming the transport. I think though that otherwise I'm ready for review/merge. The checklist of things is addressed and I've added tests and removed stdio references.

leehack commented on example/iostream-client-server/simple.dart:
```diff
@@ -0,0 +1,56 @@
+  final availableTools = await client.listTools();
+  final toolNames = availableTools.tools.map((e) => e.name);
+  print("Client setup complete. Available tools: $toolNames");
+}
```
nit: add a newline

leehack commented on example/iostream-client-server/withPipeHelper.dart:
```diff
@@ -0,0 +1,70 @@
+  final availableTools = await client.listTools();
+  final toolNames = availableTools.tools.map((e) => e.name);
+  print("Client setup complete. Available tools: $toolNames");
+}
```
nit: add a newline

leehack commented:
PR approved with nit comments.

jhgaylor commented:
Nits addressed.
